### DISCUSSION TITLE
Additional data types to support faster inference via quantization

### DISCUSSION THREAD

**Discussion by phlogisticfugu** [modular/modular#953] [Ideas]
The current state of the art in quantization https://arxiv.org/abs/2306.00978 - explores the usage of int3 and int4.

This idea is to expand the current [data types in Mojo](https://docs.modular.com/mojo/stdlib/builtin/dtype.html) to include more "exotic" low-bit data types - to enable research on improved inference via quantization.

Some of these data types may not yet be supported on hardware devices, but even if they have to fall back to higher-bit representations, there are gains to be had in model size.

**Comment by phlogisticfugu** [comment]
more on this: https://arxiv.org/abs/2301.12017

**Comment by rarebreed** [comment]
I believe that is what mojo's  MLIR interface is all about.  If I am understanding it correctly, then you should be able to write low-level MLIR code to have lower precision types as they show in [creating your own Boolean type notebook example](https://docs.modular.com/mojo/notebooks/BoolMLIR.html)

> Each of MLIR’s dialects can interoperate with the others. This is why MLIR is said to unlock heterogeneous compute: as newer, faster processors and architectures are developed, new MLIR dialects are implemented to generate optimal code for those environments. Any new MLIR dialect can be translated seamlessly into other dialects, so as more get added, all existing MLIR becomes more powerful.
>
> This means that our own custom types, such as the OurBool type we’ll create below, can be used to provide programmers with a high-level, Python-like interface. But “under the covers,” Mojo and MLIR will optimize our convenient, high-level types for each new processor that appears in the future.

**Reply by phlogisticfugu** [reply]
Thanks for responding @rarebreed .

I hadn't been aware of the MLIR interface, but my reading of it is that MLIR is for defining higher-level types (like tensors and graphs) and not more primitive types?

MLIR already seems to support low-precision integers https://mlir.llvm.org/docs/Dialects/Builtin/#integertype - so maybe it just a matter of making them more explicitly available in Mojo?  That way it'd be easier to do things like define a tensor with values of type int4.

**Reply by rarebreed** [reply]
My understanding is that MLIR allows you to create your own data types at what is essentially the compiler level (or rather, the MLIR layer).

I've been trying to wrap my head a bit around MLIR, but my understanding so far, is it is a compiler framework, that allows all kinds of exotic hardware vendors to more easily create the data types of the actual machine level ISA that each new hardware has.  These can then be contained in MLIR _dialects_.

The low-level MLIR interface that Mojo has lets you create new data types on the fly.  So I don't believe these are "high level" data types.  From the code example they give, under the covers, MLIR will ultimately generate data types that can be passed directly to registers.   From the docs (bold is my emphasis)

> Going one step further, Mojo’s unique combination of zero-cost abstractions with MLIR interoperability means that Mojo programs can take full advantage of anything that interfaces with MLIR. While this isn’t something normal Mojo programmers may ever need to do, it’s an extremely powerful capability when extending a system to interface with a new datatype, or _**an esoteric new accelerator feature.**_

The code example uses @register_passable which is how you can specify a datatype to go directly to a register (that being said, I still don't know how mojo let's you choose _which_ register it will go to?  What if there is more than one applicable register that can hold a data value?)

```python
@register_passable("trivial")
struct OurBool:
    var value: __mlir_type.i1

...
```

And way down at the bottom, this also leads me to think MLIR ultimately gets to the hardware level with zero-cost:

> As we’ve seen, Mojo’s integration with MLIR allows Mojo programmers to implement zero-cost abstractons on par with Mojo’s own _**builtin and standard library types.**_

**Comment by phlogisticfugu** [comment]
In an ideal situation, someone working on a new quantization implementation wouldn't have to write the new types themselves.

relevant for this discussion: https://www.opencompute.org/blog/amd-arm-intel-meta-microsoft-nvidia-and-qualcomm-standardize-next-generation-narrow-precision-data-formats-for-ai

there's a start at standardization on this from the chip makers
